File: src/commonTest/kotlin/com/xebia/functional/chains/SimpleSequentialChainSpec.kt
package com.xebia.functional.chains

import arrow.core.Either
import arrow.core.raise.either
import io.kotest.assertions.arrow.core.shouldBeLeft
import io.kotest.assertions.arrow.core.shouldBeRight
import io.kotest.core.spec.style.StringSpec

class SimpleSequentialChainSpec : StringSpec({
    "SimpleSequenceChain should return a prediction with one chain" {
        val chain1 = FakeChain(inputVariables = setOf("foo"), outputVariables = setOf("bar"))
        val chains = listOf(chain1)

        either {
            val ssc = SimpleSequentialChain(chains = chains, returnAll = true)
            ssc.run(mapOf("input" to "123")).bind()
        } shouldBeRight mapOf("input" to "123", "output" to "123dr")
    }

    "SimpleSequenceChain should return a prediction with more than one chain" {
        val chain1 = FakeChain(inputVariables = setOf("foo"), outputVariables = setOf("bar"))
        val chain2 = FakeChain(inputVariables = setOf("bar"), outputVariables = setOf("baz"))
        val chain3 = FakeChain(inputVariables = setOf("baz"), outputVariables = setOf("dre"))
        val chains = listOf(chain1, chain2, chain3)

        either {
            val ssc = SimpleSequentialChain(chains = chains, returnAll = true)
            ssc.run(mapOf("input" to "123")).bind()
        } shouldBeRight mapOf("input" to "123", "output" to "123drdrdr")
    }

    "SimpleSequentialChain should fail if multiple input variables are expected" {
        val chain1 = FakeChain(inputVariables = setOf("foo"), outputVariables = setOf("bar"))
        val chain2 = FakeChain(inputVariables = setOf("bar", "foo"), outputVariables = setOf("baz"))
        val chains = listOf(chain1, chain2)

        either {
            val ssc = SimpleSequentialChain(chains = chains, returnAll = true)
            ssc.run(mapOf("input" to "123")).bind()
        } shouldBeLeft SequenceChain.InvalidKeys("The expected inputs are more than one: {bar}, {foo}")
    }

    "SimpleSequentialChain should fail if multiple output variables are expected" {
        val chain1 = FakeChain(inputVariables = setOf("foo"), outputVariables = setOf("bar", "foo"))
        val chain2 = FakeChain(inputVariables = setOf("bar"), outputVariables = setOf("baz"))
        val chains = listOf(chain1, chain2)

        either {
            val ssc = SimpleSequentialChain(chains = chains, returnAll = true)
            ssc.run(mapOf("input" to "123")).bind()
        } shouldBeLeft SequenceChain.InvalidKeys("The expected outputs are more than one: {bar}, {foo}")
    }
})

data class FakeChain(private val inputVariables: Set<String>, private val outputVariables: Set<String>) : Chain {
    override val config: Chain.Config = Chain.Config(
        inputKeys = inputVariables,
        outputKeys = outputVariables,
        returnAll = false
    )

    override suspend fun call(inputs: Map<String, String>): Either<Chain.InvalidInputs, Map<String, String>> =
        either {
            val variables = inputVariables.map { inputs[it] }.requireNoNulls()
            outputVariables.fold(emptyMap()) { outputs, outputVar ->
                outputs + (outputVar to "${variables.joinToString(separator = "")}dr")
            }
        }
}

File: scala/src/test/scala/com/xebia/functional/chains/SimpleSequentialChainSpec.scala
package com.xebia.functional.scala.chains

import cats.data.NonEmptySeq
import cats.effect.IO
import cats.implicits.*

import com.xebia.functional.scala.chains.mock.FakeChain
import com.xebia.functional.scala.chains.models.*
import eu.timepit.refined.types.string.NonEmptyString
import munit.CatsEffectSuite

class SimpleSequentialChainSpec extends CatsEffectSuite:
  test("SimpleSequentialChain functionality") {
    val chain1 = FakeChain(inputVariables = Set("foo"), outputVariables = Set("bar"))
    val chain2 = FakeChain(inputVariables = Set("bar"), outputVariables = Set("baz"))
    val chains = NonEmptySeq(chain1, Seq(chain2))

    val inputKey = NonEmptyString.unsafeFrom("input")
    val outputKey = NonEmptyString.unsafeFrom("output")

    val ssc = SimpleSequentialChain.resource[IO](chains, inputKey, outputKey)

    val output = ssc.use(_.run(Map("input" -> "123")))
    val expectedOutput = Map("output" -> "123foofoo", "input" -> "123")
    assertIO(output, expectedOutput)
  }

  test("SimpleSequentialChain should fail if multiple input variables are expected") {
    val chain1 = FakeChain(inputVariables = Set("foo"), outputVariables = Set("bar"))
    val chain2 = FakeChain(inputVariables = Set("bar", "foo"), outputVariables = Set("baz"))
    val chains = NonEmptySeq(chain1, Seq(chain2))

    val inputKey = NonEmptyString.unsafeFrom("input")
    val outputKey = NonEmptyString.unsafeFrom("output")

    val ssc = SimpleSequentialChain.resource[IO](chains, inputKey, outputKey)

    val output = ssc.use(_.run(Map("input" -> "123")))

    interceptIO[InvalidChainInputError](output)
  }

  test("SimpleSequentialChain should fail if multiple outputs variables are expected") {
    val chain1 = FakeChain(inputVariables = Set("foo"), outputVariables = Set("bar", "becue"))
    val chain2 = FakeChain(inputVariables = Set("bar"), outputVariables = Set("baz"))
    val chains = NonEmptySeq(chain1, Seq(chain2))

    val inputKey = NonEmptyString.unsafeFrom("input")
    val outputKey = NonEmptyString.unsafeFrom("output")

    val ssc = SimpleSequentialChain.resource[IO](chains, inputKey, outputKey)

    val output = ssc.use(_.run(Map("input" -> "123")))

    interceptIO[InvalidChainOutputError](output)
  }
